### ISSUE TITLE
Digital Sequencer probability

### ISSUE THREAD

**Issue by matthleg** [clone45/voxglitch#74]
Hi, first thanks a lot for your great modules, Digital Sequencer is the sequencer I use the most: easy to use, very visual and lot of possibilities.
To do it really perfect I would like to have probability for each step ( like in the image?); I think it would be the only sequencer in vcv with this possibility... 
![digital sequencer probability](https://user-images.githubusercontent.com/36376831/115112517-2f201d80-9f86-11eb-89e1-b0682c34b7a7.jpg)

Please think about it, I know it's lot of work to make (or change) module, so again thanks a lot.
All the best,
                  Matthieu.

**Comment by clone45** [comment]
Hi Matthieu,

I love this idea, but it would probably be best as a expansion module.  I'll see what I can do!  I haven't been working on this module for a while, and I realize that there are a few outstanding requests.  My apologies.  Sometime life gets in the way. :-)

**Comment by matthleg** [comment]
Days are too short... I know this, but I am happy you like the idea. It
would be great if you manage to do it one day.

**Comment by clone45** [comment]
Here's an idea that I had to pull it off.  I could create a small icon at the left or right hand side.  When clicked on, it'll slide out a separate "probability" sequencer similar to the one you designed.  What do you think?

![ds-pcent](https://user-images.githubusercontent.com/1373491/115156813-03ab3a80-a03b-11eb-9c24-aed23dfc9d56.png)

**Comment by matthleg** [comment]
I think it would be perfect!

**Comment by S0210** [comment]
I don't think there's a need for a switch. We could have the current gates as well as the probability CV influenced gates: 
![image](https://user-images.githubusercontent.com/34127628/133004075-18db8ea6-ddcd-4cc5-acf4-28d4e1494efb.png)

OK, and now let me have a question that is related to the picture:
* step 1 gate will be sent for sure (100%) 
* step 5 gate will be most likely sent (ca. 85%)
* what about step 4? (I'd love to have a slight chance that the pulse is sent)
Is it the behavior you imagined?

Or was this idea dropped completely? 

Regards, Andras

**Comment by clone45** [comment]
Hi Andras,

I really like your concept, but I haven't had the time to focus on module building for a while.  I have an idea on how to pull this off, so let me set this coming Wednesday afternoon aside to give it a shot. :-)  How important is CV control over the gate percentage to you?  Would it be adequate if it was, for example, a knob?

**Comment by S0210** [comment]
Hello,
Just think about it. It's always more complicated than it looks. By the way, those %GATE sockets on my drawing are **outputs**. Sorry for the confusion. :) 
I don't see myself using CV control over the gate percentage at all.   
Best regards, Andras

**Comment by clone45** [comment]
I wonder, could a separate 8x probability gate sequencer serve the same purpose?  That wouldn't be very difficult to build, and although it might take a bit more space in a patch, it could be useful independently of the Digital Sequencer.

**Comment by S0210** [comment]
The original suggestion by matthleg was about a per step probability. I don't see how an external module could serve that purpose. 

Having a separate 8x probability gate module (with input & knob & output) sounds like a wonderful idea. I created a prototype but [change extension to .vcv](https://orokos.neocities.org/temp/VCV%20Trigger%20Precision%20-%20rename%20to%20vcv.txt) after downloading. It creates a lot of movement in rhythms and I absolutely love it. I tried to use K1 knob to select if the script only removes, or only adds, or does both.  

![image](https://user-images.githubusercontent.com/34127628/133238443-4e0bf2f2-4807-4ec7-acd5-38b0a7cb5c0c.png)

Ugh, I think I reached the limits of my coding skills. So did I with my artistic skills with designing a layout proposal:

![image](https://user-images.githubusercontent.com/34127628/133241330-2bb607aa-096a-407b-8c8e-dea0e29fd55c.png)

Have a wonderful Wednesday afternoon! :)

Cheers, Andras

**Comment by matthleg** [comment]
Yes you're right, my interest is more focused on the per step probability, but I am not programmer so I have no idea of the difficulties... Sent from my Huawei Mobile-------- Original message --------From: Andras Szabo ***@***.***>Date: Tue, 14 Sep 2021, 12:33To: clone45/voxglitch ***@***.***>Cc: matthleg ***@***.***>, Author ***@***.***>Subject: Re: [clone45/voxglitch] Digital Sequencer probability (#74)
The original suggestion by matthleg was about a per step probability. I don't see how an external module could serve that purpose.
Having a separate 8x probability gate module (with input & knob & output) sounds like a wonderful idea. I created a prototype but change extension to .vcv after downloading. It creates a lot of movement in rhythms and I absolutely love it. I tried to use K1 knob to select if the script only removes, or only adds, or does both.

Ugh, I think I reached the limits of my coding skills. So did I with my artistic skills with designing a layout proposal:

Have a wonderful Wednesday afternoon! :)
Cheers, Andras

**Comment by S0210** [comment]
Hoping that the external Trigger Precision module makes it to Bret's list of future projects, I'd like to comment on the original 'per step probability' proposal.

For @matthleg the great [StochSeq4](https://library.vcvrack.com/Sha-Bang-Modules/StochSeq4) might be a good solution. It is visually less appealing but it's an excellent piece of work. It's specifically built for per step probability. I don't know if it's worth the hassle to re-create this feature in Digital Sequencer.

**Comment by S0210** [comment]
This thread was de-routed by the Trigger Precision plans.

However I'd like to comment on @matthleg 's  per-step-probability proposal once again. The soon to be issued [RndG ](https://library.vcvrack.com/dbRackModules/RndG) v2.0.5 might be the perfect solution. 

It accepts the GATE of sequence #1 and will modify the probability by CV of sequence #1 (-5V to +5V). It is more flexible than a built-in Digital Sequencer feature as you can decide to use the same CV for multiple GATE outputs. Or even link a MIDI controller to the probability input and change it manually on the fly.

<img width="772" alt="image" src="https://user-images.githubusercontent.com/34127628/150132301-bdade521-2ef9-444e-89ad-0916fd2fd9d7.png">

I'm afraid it's just one more argument _against a built-in_ per-step-probability in Digital Sequencer.

**Comment by clone45** [comment]
@S0210 Andras, thanks for pointing that out.
